Dev instructions for new releases

choldgraf opened the pull request:
This adds more fleshed out instructions for cutting a new release, and was inspired by the most recent (v0.6) release of JupyterHub's helm chart. Somebody other than @choldgraf should test this out the next time we need to make a release!

choldgraf commented:
I added a checklist that people can copy/paste into issues to manage this process!

willingc commented on CONTRIBUTING.md:
```diff
@@ -54,19 +54,189 @@ development.
+## Releasing a new version of the helm chart
+
+The following steps can be followed to release a new version of the Helm Chart.
+This should happen approximately once every 5-7 weeks.
```
Presently, we expect a release approximately every 5-7 weeks.

willingc commented on CONTRIBUTING.md:
```diff
@@ -54,19 +54,189 @@ development.
+- [ ] Make a CHANGELOG
+- [ ] Generate and add the list of contributors
+- [ ] Build and push a new Docker image to DockerHub
+- [ ] Commit version bump in `Chartyaml` and `Values.yaml`
```
Chart.yaml

willingc commented on CONTRIBUTING.md:
```diff
@@ -54,19 +54,189 @@ development.
+- [ ] Create and push a new tag for this release
+- [ ] Create and publish a new GitHub release
+- [ ] Write / publish a blog post based largely off of the CHANGELOG
+- [ ] Set ReadTheDocs to begin using `latest` by default
```
Is there a reason for this?

willingc replied:
Ah nevermind. I see why a few lines down.

choldgraf replied:
think we could get away with just having docs that point to the last version, then updating straight from `last-version` to `new-version` instead of going through `latest`?

willingc replied:
I'm inclined to just leave enough time before release (i.e. RC) to have the docs ready at the same point as the release of the charts. If the docs are radically off, then there should be a dot release perhaps. I'll defer to you and @yuvipanda but it would be great to release all at once as any minor docs changes will be available in latest anyway.

yuvipanda replied:
Yeah, I think we can do this once enough of our docs get stable enough, but I think for now, we should keep this unfortunately. For example, we run into problems with GKE version needing to be bumped quite often still, so if we don't switch to latest users won't be able to even complete a setup

willingc commented on CONTRIBUTING.md:
```diff
@@ -54,19 +54,189 @@ development.
+* [JupyterHub](https://github.com/jupyterhub/jupyterhub)
+* [OAuthenticator](https://github.com/jupyterhub/oauthenticator))
+
+edit `contributors.py` to have the appropriate dates
```
Edit

choldgraf commented:
ok, latest push addresses comments + tries to clarify a bit how the documentation should be handled with new releases. LMKWYT!

willingc commented on CONTRIBUTING.md:
```diff
@@ -54,19 +54,198 @@ development.
+chart, {{release-name}}. Instructions for creating the release can be found in
+[CONTRIBUTING.md](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/master/CONTRIBUTING.md#releasing-a-new-version-of-the-helm-chart).
+Below is the checklist for this release.
+
```
I would like to an explicit step 0: Code, tests, and documentation to support a release are stable.

willingc replied:
I'm firm on wanting this explicitly stated. This should be the goal even if it takes a release or two to get there.

choldgraf replied:
Definitely agree w/ you on this point 👍 my question only pertained to the other comment

willingc commented on CONTRIBUTING.md:
```diff
@@ -54,19 +54,198 @@ development.
+- [ ] Generate and add the list of contributors
+- [ ] Build and push a new Docker image to DockerHub
+- [ ] Commit version bump in `Chart.yaml` and `Values.yaml`
+- [ ] Update references in documentation to the new version
```
Update references in documentation to the new version (note: documentation should be stable and there should be no anticipated major changes to content).

willingc commented on CONTRIBUTING.md:
```diff
@@ -54,19 +54,198 @@ development.
+
+### Extra step - release a documentation release
+
+It is common that documentation changes are made shortly after a new release.
```
I'm inclined to say let's just do this as part of the release process (making the time window: 2 weeks or when the next release docs begin to be written).

Otherwise, we are generating more work for us and more confusion for the end user (a large portion will looks for the current version docs).

choldgraf replied:
IIRC the reasoning was that there are still a lot of large-ish changes to the docs happening, so it'd be best if the docs always show the latest updates until there's something written that doesn't apply to the currently-released helm chart. 

What if we kept this behavior for one or two release cycles, and then once the docs stabilize we can start to just jump straight from `last-version` to `next-version`?

cc @yuvipanda in case you've got thoughts as well

willingc replied:
As long as the comment (Code, tests, and documentation to support a release are stable.) goes in as the first checklist step, I'm fine doing this as you propose for a couple of releases.

choldgraf commented:
@willingc what do you think about the latest push? I added a note at the end suggesting that this was temporary behavior, and added a first step of "stable code / docs / etc" to the checklist

choldgraf commented:
@yuvipanda wanna push the green button on this if you're 👍 on the language?

yuvipanda commented:
Thanks for making this happen, @choldgraf!
